istioctl/pkg/multicluster/remote_secret.go

```diff
@@ -0,0 +1,270 @@
+	flags.StringVar(&o.ServiceAccountName, "service-account-name", o.ServiceAccountName,
+		"name of Pilot service account in the remote cluster(s)")
+
+	return c
+}
+
+// hooks for testing
+var (
+	newStartingConfig = func(kubeconfig string) (*api.Config, error) {
+		return kube.BuildClientCmd(kubeconfig, "").ConfigAccess().GetStartingConfig()
+	}
+
+	newKubernetesInterface = func(kubeconfig, context string) (kubernetes.Interface, error) {
+		return kube.CreateClientset(kubeconfig, context)
+	}
+)
+
+const (
+	caDataSecretKey = "ca.crt"
+	tokenSecretKey  = "token"
+)
+
+func createRemotePilotKubeconfig(in *v1.Secret, config *api.Config, context string) (*api.Config, error) {
+	contextInfo, ok := config.Contexts[context]
+	if !ok {
+		return nil, fmt.Errorf("%q context not found in Kubeconfig(s)", context)
+	}
+
+	clusterInfo, ok := config.Clusters[contextInfo.Cluster]
+	if !ok {
+		return nil, fmt.Errorf("%q cluster not found in Kubeconfig(s) for context %q", contextInfo.Cluster, context)
+	}
+
+	caData, ok := in.Data[caDataSecretKey]
+	if !ok {
+		return nil, fmt.Errorf("no %q data found in secret %s/%s for context %q",
+			caDataSecretKey, in.Namespace, in.Name, context)
+	}
+
+	token, ok := in.Data[tokenSecretKey]
+	if !ok {
+		return nil, fmt.Errorf("no %q data found in secret %s/%s for context %q",
+			tokenSecretKey, in.Namespace, in.Name, context)
+	}
+
+	kubeconfig := &api.Config{
+		Clusters: map[string]*api.Cluster{
+			contextInfo.Cluster: {
+				CertificateAuthorityData: caData,
+				Server:                   clusterInfo.Server,
+			},
+		},
+		AuthInfos: map[string]*api.AuthInfo{
+			contextInfo.Cluster: {
+				Token: string(token),
+			},
+		},
+		Contexts: map[string]*api.Context{
+			contextInfo.Cluster: {
+				Cluster:  contextInfo.Cluster,
+				AuthInfo: contextInfo.Cluster,
+			},
+		},
+		CurrentContext: contextInfo.Cluster,
+	}
+	return kubeconfig, nil
+}
+
+func createRemotePilotServiceAccountSecret(kubeconfig *api.Config, name string) (*v1.Secret, error) { // nolint:interfacer
+	var data bytes.Buffer
+	if err := latest.Codec.Encode(kubeconfig, &data); err != nil {
+		return nil, err
+	}
+	out := &v1.Secret{
+		ObjectMeta: metav1.ObjectMeta{
+			Name: fmt.Sprintf("%v%v", defaultSecretPrefix, name),
+			Labels: map[string]string{
+				secretcontroller.MultiClusterSecretLabel: "true",
+			},
+		},
+		StringData: map[string]string{
+			name: data.String(),
+		},
+	}
+	return out, nil
+}
+
+func createPilotRemoteSecret(out *bytes.Buffer, config *api.Config, name, context string, o *Options) error {
+	if _, err := out.WriteString(fmt.Sprintf("# Remote pilot credentials for cluster context %q\n", context)); err != nil {
+		return err
+	}
+
+	kube, err := newKubernetesInterface(o.Kubeconfig, context)
+	if err != nil {
+		return err
+	}
+
+	// Get the remote pilot's service-account-token secret
+	serviceAccount, err := kube.CoreV1().ServiceAccounts(o.Namespace).Get(o.ServiceAccountName, metav1.GetOptions{})
+	if err != nil {
```

Comment:
This is a great start - but maybe add a TODO to extend this to use certs ( some example code for creating k8s-signed certs in https://github.com/costinm/webhook-certificate-generator ), and maybe other forms of k8s auth.

Reply:
Good point - added a reference to TODO referencing https://kubernetes.io/docs/reference/access-authn-authz/authentication/#authentication-strategies